posts/2023-10-05-Rust-1.73.0.md

```diff
@@ -0,0 +1,114 @@
+
+## What's in 1.73.0 stable
+
+## Cleaner panic messages
+
+The output produced by the default panic handler has been changed
+to put the panic message on its own line instead of wrapping it in quotes.
+This can make panic messages easier to read, as shown in this example:
+
+<div style="margin:1em"><pre><code>fn main() {
+    let file = "ferris.txt";
+    panic!("oh no! {file:?} not found!");
+}</code></pre>
+Output before Rust 1.73:
+<pre style="margin-top:0"><code style="background:#000;color:#ccc" class="language-text">thread 'main' panicked at 'oh no! "ferris.txt" not found!', src/main.rs:3:5</code></pre>
+Output starting in Rust 1.73:
+<pre style="margin-top:0"><code style="background:#000;color:#ccc" class="language-text">thread 'main' panicked at src/main.rs:3:5:
+oh no! "ferris.txt" not found!</code></pre></div>
+
+This is especially useful when the message is long, contains nested quotes, or spans multiple lines.
+
+Additionally, the panic messages produced by `assert_eq` and `assert_ne` have
+been modified, moving the custom message (the third argument)
+and removing some unnecessary punctuation, as shown below:
+
+<div style="margin:1em"><pre><code>fn main() {
+    assert_eq!("🦀", "🐟", "ferris is not a fish");
+}</code></pre>
+Output before Rust 1.73:
+<pre style="margin-top:0"><code style="background:#000;color:#ccc" class="language-text">thread 'main' panicked at 'assertion failed: `(left == right)`
+ left: `"🦀"`,
+right: `"🐟"`: ferris is not a fish', src/main.rs:2:5</code></pre>
+Output starting in Rust 1.73:
+<pre style="margin-top:0"><code style="background:#000;color:#ccc" class="language-text">thread 'main' panicked at src/main.rs:2:5:
+assertion `left == right` failed: ferris is not a fish
+ left: "🦀"
+right: "🐟"</code></pre></div>
+
+### Thread local initialization
+
+As proposed in [RFC 3184](https://github.com/rust-lang/rfcs/blob/master/text/3184-thread-local-cell-methods.md), `LocalKey<Cell<T>>` and `LocalKey<RefCell<T>>` can now be directly manipulated with `get()`, `set()`, `take()`, and `replace()` methods, rather than jumping through a `with(|inner| ...)` closure as needed for general `LocalKey` work. `LocalKey<T>` is the type of `thread_local!` statics.
+
+The new methods make common code more concise and avoid running the extra initialization code for the default value specified in `thread_local!` for new threads.
+
+```rust
+thread_local! {
+    static THINGS: RefCell<Vec<i32>> = RefCell::new(Vec::new());
+}
+
+fn f() {
+    // before:
+    THINGS.with(|i| i.set(vec![1, 2, 3]));
+    // now:
+    THINGS.set(vec![1, 2, 3]);
+
+    // ...
+
+    // before:
+    let v = THINGS.with(|i| i.take());
+    // now:
+    let v: Vec<i32> = THINGS.take();
+}
```

Comment:
Is this example correct? When I try to run the "before" code in Rust 1.72.1, I get a compiler error. Here's my `main.rs` file:

```rust
use std::cell::RefCell;

thread_local! {
    static THINGS: RefCell<Vec<i32>> = RefCell::new(Vec::new());
}

fn main() {
    THINGS.with(|i| i.set(vec![1, 2, 3]));
    let v = THINGS.with(|i| i.take());
}
```

and here's the error:

```text
$ cargo run
   Compiling scratch v0.1.0 (/Users/nathan/scratch)
error[E0599]: no method named `set` found for reference `&RefCell<Vec<i32>>` in the current scope
   --> src/main.rs:8:23
    |
8   |     THINGS.with(|i| i.set(vec![1, 2, 3]));
    |                       ^^^ method not found in `&RefCell<Vec<i32>>`
    |
   ::: /Users/nathan/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/pin.rs:747:12
    |
747 |     pub fn set(&mut self, value: P::Target)
    |            --- the method is available for `Pin<&mut &RefCell<Vec<i32>>>` here
    |
help: consider wrapping the receiver expression with the appropriate type
    |
8   |     THINGS.with(|i| Pin::new(&mut i).set(vec![1, 2, 3]));
    |                     +++++++++++++  +

For more information about this error, try `rustc --explain E0599`.
error: could not compile `scratch` (bin "scratch") due to previous error
```

If I change `RefCell` to `Cell`, then the code compiles. Maybe that was what was intended for the example?

```suggestion
thread_local! {
    static THINGS: Cell<Vec<i32>> = Cell::new(Vec::new());
}

fn f() {
    // before:
    THINGS.with(|i| i.set(vec![1, 2, 3]));
    // now:
    THINGS.set(vec![1, 2, 3]);

    // ...

    // before:
    let v = THINGS.with(|i| i.take());
    // now:
    let v: Vec<i32> = THINGS.take();
}
```

Reply:
Seems plausible - I don't think I tried to check the example at all.

Reply:
Ah, the `LocalKey<T>` method `set` for both `T: RefCell` and `T: Cell`  is mimicking the API of `Cell`'s `set` method, but `RefCell` doesn't have that method.

I think changing this to a `Cell` example was a good choice, as the example is easy to understand. If it were to remain a `RefCell` example, it would need to be changed something like the code below.

```rust
// before:
THINGS.with(|i| *i.borrow_mut() = vec![1, 2, 3]);
// now:
THINGS.set(vec![1, 2, 3]);
```